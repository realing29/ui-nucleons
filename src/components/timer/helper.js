import parseISO from 'date-fns/parseISO';
import isAfter from 'date-fns/isAfter';
import formatDistanceToNowStrict from 'date-fns/formatDistanceToNowStrict';
import padCharsStart from 'lodash/fp/padCharsStart';

const UNITS = ['day', 'hour', 'minute', 'second'];

const padTimePart = padCharsStart('0', 2);

/**
 * Возвращает отформатированную разницу во времени между событием и текущим временем.
 * @param {string} dateString Дата и время события в формате ISO.
 * @return {string} Отформатированное время до текущего момента в формате d:hh:mm:ss.
 */
export const getTimeDurationToNow = dateString => {
  const date = parseISO(dateString);
  let result;

  if (isAfter(new Date(), date)) {
    result = '0:00:00:00';
  } else {
    const [days, hours, minutes, seconds] = UNITS.map(
      unit => formatDistanceToNowStrict(date, {
        unit,
        roundingMethod: 'floor',
      }).replace(/\D/g, '') // результат содержит слова - убираем их
    );

    result = [
      days,
      ...[hours % 24, minutes % 60, seconds % 60].map(padTimePart), // вместо "2" выводим "02"
    ].join(':');
  }

  return result;
};
