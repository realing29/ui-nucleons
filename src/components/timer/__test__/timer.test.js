import { mount } from 'enzyme';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import React from 'react';
import Timer from '../index';
import { getTimeDurationToNow } from '../helper';

jest.useFakeTimers();

jest.mock('../helper', () => {
  const original = jest.requireActual('../helper');
  return {
    ...original,
    __esModule: true,
    getTimeDurationToNow: jest.fn(),
  };
});

beforeEach(() => {
  jest.clearAllTimers();
  jest.clearAllMocks();
});

describe('Timer', () => {
  it('should render properly without props', () => {
    const wrapper = mount(
      <Timer />
    );
    expect(wrapper.find('time')).toHaveLength(1);
  });

  it('should pass props', () => {
    const testStyle = {
      className: 'timer',
      style: {
        padding: 10,
        margin: 10,
        background: '#eee',
      },
    };
    const wrapper = mount(
      <Timer
        endTime='2020-03-08'
        format='D ะด., hh:mm:ss'
        timeProps={testStyle}
      />
    );
    expect(wrapper.find('time').at(0).prop('className')).toBe('timer');
    expect(wrapper.find('time').at(0).prop('style')).toEqual(testStyle.style);
  });
  it('should change time by timeout', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);

    act(() => {
      render(<Timer endTime='30' />, container);
    });

    expect(getTimeDurationToNow).toHaveBeenCalledTimes(1);
    act(() => {
      jest.advanceTimersByTime(5000);
    });
    expect(getTimeDurationToNow).toHaveBeenCalledTimes(6);
  });

  it('should reset interval on unmount', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    act(() => {
      render(<Timer endTime='30' />, container);
    });
    expect(clearInterval).toHaveBeenCalledTimes(0);
    act(() => {
      unmountComponentAtNode(container);
    });
    expect(clearInterval).toHaveBeenCalledTimes(1);
  });
});
