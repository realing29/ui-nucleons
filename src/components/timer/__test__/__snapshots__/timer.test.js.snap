// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer should handle "calculate" prop 1`] = `
<Timer
  calculate={[Function]}
  endTime="2020-01-01 22:00"
>
  <time>
    2020-01-01 22:00
  </time>
</Timer>
`;
